crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1132 @@
+                system_path_to_file(db.upcast(), path)
+            }
+            (Self::StandardLibrary(path), Self::StandardLibrary(stdlib_root)) => {
+                match Self::query_stdlib_version(path, typeshed_versions, search_path, stdlib_root, db, target_version) {
+                    TypeshedVersionsQueryResult::DoesNotExist => None,
+                    TypeshedVersionsQueryResult::Exists => system_path_to_file(db.upcast(), path),
+                    TypeshedVersionsQueryResult::MaybeExists => system_path_to_file(db.upcast(), path)
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    fn to_module_name(self) -> Option<ModuleName> {
+        let (fs_path, skip_final_part) = match self {
+            Self::Extra(path) | Self::FirstParty(path) | Self::SitePackages(path) => (
+                path,
+                path.ends_with("__init__.py") || path.ends_with("__init__.pyi"),
+            ),
+            Self::StandardLibrary(path) => (path, path.ends_with("__init__.pyi")),
+        };
+
+        let parent_components = fs_path
+            .parent()?
+            .components()
+            .map(|component| component.as_str());
+
+        if skip_final_part {
+            ModuleName::from_components(parent_components)
+        } else {
+            ModuleName::from_components(parent_components.chain(fs_path.file_stem()))
+        }
+    }
+
+    #[must_use]
+    fn with_pyi_extension(&self) -> ModuleResolutionPathBufInner {
+        match self {
+            Self::Extra(path) => ModuleResolutionPathBufInner::Extra(path.with_extension("pyi")),
+            Self::FirstParty(path) => {
+                ModuleResolutionPathBufInner::FirstParty(path.with_extension("pyi"))
+            }
+            Self::StandardLibrary(path) => {
+                ModuleResolutionPathBufInner::StandardLibrary(path.with_extension("pyi"))
+            }
+            Self::SitePackages(path) => {
+                ModuleResolutionPathBufInner::SitePackages(path.with_extension("pyi"))
+            }
+        }
+    }
+
+    #[must_use]
+    fn with_py_extension(&self) -> Option<ModuleResolutionPathBufInner> {
+        match self {
+            Self::Extra(path) => Some(ModuleResolutionPathBufInner::Extra(
+                path.with_extension("py"),
+            )),
+            Self::FirstParty(path) => Some(ModuleResolutionPathBufInner::FirstParty(
+                path.with_extension("py"),
+            )),
+            Self::StandardLibrary(_) => None,
+            Self::SitePackages(path) => Some(ModuleResolutionPathBufInner::SitePackages(
+                path.with_extension("py"),
+            )),
+        }
+    }
+
+    #[must_use]
+    fn relativize_path(&self, absolute_path: &'a FileSystemPath) -> Option<Self> {
```

Comment:
I'm finding it quite hard to understand the handling of relative vs absolute paths. Here in this method we create instances of `ModuleResolutionPathRef` that contain relative paths, but then in the above methods it seems like we expect them to contain absolute paths, because for non-stdlib variants we just ignore the passed-in search path entirely.

I think we should establish some clarity/invariants around which kinds of paths should be absolute and which should be relative.